# The second task for 'sports.ru'
### Andrey Shubin

In [1]:
#at first it's necessary to import librares:
#pandas for work with dataframe and scipy for statistics functions

import pandas as pd
from scipy import stats

In [2]:
#loading dataset to pandas dataframe 'df'

df = pd.read_csv('dataset_Facebook.csv', sep = ';')

In [3]:
#starting aggregation of all dataset with describe method
#it's possible to check mean (the average value), minimum, maximum, values of 25th, 50th, and 75th percentiles
#for each column

df.describe()

Page total likes    Category  Post Month  Post Weekday   Post Hour  \
count        500.000000  500.000000  500.000000    500.000000  500.000000   
mean      123194.176000    1.880000    7.038000      4.150000    7.840000   
std        16272.813214    0.852675    3.307936      2.030701    4.368589   
min        81370.000000    1.000000    1.000000      1.000000    1.000000   
25%       112676.000000    1.000000    4.000000      2.000000    3.000000   
50%       129600.000000    2.000000    7.000000      4.000000    9.000000   
75%       136393.000000    3.000000   10.000000      6.000000   11.000000   
max       139441.000000    3.000000   12.000000      7.000000   23.000000   

             Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
count  499.000000                  500.00000                     5.000000e+02   
mean     0.278557                13903.36000                     2.958595e+04   
std      0.448739                22740.78789                     7.680325e+04   
min      0.000000                  238.00000                     5.700000e+02   
25%      0.000000                 3315.00000                     5.694750e+03   
50%      0.000000                 5281.00000                     9.051000e+03   
75%      1.000000                13168.00000                     2.208550e+04   
max      1.000000               180480.00000                     1.110282e+06   

       Lifetime Engaged Users  Lifetime Post Consumers  \
count              500.000000               500.000000   
mean               920.344000               798.772000   
std                985.016636               882.505013   
min                  9.000000                 9.000000   
25%                393.750000               332.500000   
50%                625.500000               551.500000   
75%               1062.000000               955.500000   
max              11452.000000             11328.000000   

       Lifetime Post Consumptions  \
count                  500.000000   
mean                  1415.130000   
std                   2000.594118   
min                      9.000000   
25%                    509.250000   
50%                    851.000000   
75%                   1463.000000   
max                  19779.000000   

       Lifetime Post Impressions by people who have liked your Page  \
count                                       5.000000e+02              
mean                                        1.676638e+04              
std                                         5.979102e+04              
min                                         5.670000e+02              
25%                                         3.969750e+03              
50%                                         6.255500e+03              
75%                                         1.486050e+04              
max                                         1.107833e+06              

       Lifetime Post reach by people who like your Page  \
count                                        500.000000   
mean                                        6585.488000   
std                                         7682.009405   
min                                          236.000000   
25%                                         2181.500000   
50%                                         3417.000000   
75%                                         7989.000000   
max                                        51456.000000   

       Lifetime People who have liked your Page and engaged with your post  \
count                                         500.000000                     
mean                                          609.986000                     
std                                           612.725618                     
min                                             9.000000                     
25%                                           291.000000                     
50%                                           412.000000                     
75%     

In [4]:
#we can check parameters with other methods, like

print(df.mean())

#df.median(), df.min(), df.max() and etc

Page total likes                                                       123194.176000
Category                                                                    1.880000
Post Month                                                                  7.038000
Post Weekday                                                                4.150000
Post Hour                                                                   7.840000
Paid                                                                        0.278557
Lifetime Post Total Reach                                               13903.360000
Lifetime Post Total Impressions                                         29585.948000
Lifetime Engaged Users                                                    920.344000
Lifetime Post Consumers                                                   798.772000
Lifetime Post Consumptions                                               1415.130000
Lifetime Post Impressions by people who have liked your Page     

In [5]:
#for mode using scipy function, cause built-in pandas (pd.mode()) doesn't work correctly 
#creating new dataframe and add mode and count values 

mode = pd.DataFrame(columns = ['Column name', 'Mode value', 'Count value'])
for column in df:
    mode = mode.append({'Column name':column, 'Mode value':stats.mode(df[column])[0], 'Count value':stats.mode(df[column])[1]}, ignore_index = True)

mode

D:\Progs\Anaconda3\lib\site-packages\scipy\stats\stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


Column name Mode value Count value
0                                    Page total likes   [136393]        [18]
1                                                Type    [Photo]       [426]
2                                            Category        [1]       [215]
3                                          Post Month       [10]        [60]
4                                        Post Weekday        [7]        [82]
5                                           Post Hour        [3]       [105]
6                                                Paid      [0.0]       [360]
7                           Lifetime Post Total Reach      [677]         [2]
8                     Lifetime Post Total Impressions     [4372]         [2]
9                              Lifetime Engaged Users      [537]         [4]
10                            Lifetime Post Consumers      [182]         [3]
11                         Lifetime Post Consumptions      [431]         [3]
12  Lifetime Post Impressions by people who have l...     [1210]         [2]
13   Lifetime Post reach by people who like your Page      [690]         [2]
14  Lifetime People who have liked your Page and e...      [403]         [5]
15                                            comment        [0]       [106]
16                                               like     [98.0]         [7]
17                                              share     [13.0]        [20]
18                                 Total Interactions        [0]         [6]

In [6]:
#checking list of 'Type' variables
#then find the same parameters for each object from that list

df.groupby(['Type']).groups.keys()

dict_keys(['Link', 'Photo', 'Status', 'Video'])

In [7]:
#all necessary stuff

df.groupby('Type').agg('min')
df.groupby('Type').agg('max')
df.groupby('Type').agg('mean')
df.groupby('Type').agg('median')

#in mode example pay attencion - each attribute has 2 numbers: 1 - mode value, 2 - counts value

df.groupby('Type').agg(stats.mode)

Page total likes      Category    Post Month Post Weekday  \
Type                                                                
Link     ([138353], [4])   ([1], [20])   ([12], [6])   ([6], [5])   
Photo   ([124940], [17])  ([1], [185])  ([10], [49])  ([7], [71])   
Status   ([139441], [4])   ([2], [38])  ([12], [12])   ([5], [9])   
Video    ([137893], [2])    ([1], [7])   ([11], [3])   ([2], [2])   

           Post Hour            Paid Lifetime Post Total Reach  \
Type                                                             
Link      ([3], [6])   ([0.0], [16])             ([1536], [1])   
Photo    ([3], [89])  ([0.0], [306])              ([677], [2])   
Status  ([10], [10])   ([0.0], [35])             ([3930], [1])   
Video    ([11], [3])    ([1.0], [4])            ([13544], [1])   

       Lifetime Post Total Impressions Lifetime Engaged Users  \
Type                                                            
Link                     ([3094], [1])            ([66], [2])   
Photo                    ([4372], [2])           ([537], [4])   
Status                   ([7509], [1])           ([128], [1])   
Video                   ([30235], [1])           ([459], [1])   

       Lifetime Post Consumers Lifetime Post Consumptions  \
Type                                                        
Link              ([322], [2])                ([26], [1])   
Photo             ([298], [3])               ([431], [3])   
Status             ([86], [1])              ([1692], [2])   
Video             ([411], [1])               ([539], [1])   

       Lifetime Post Impressions by people who have liked your Page  \
Type                                                                  
Link                                        ([2307], [1])             
Photo                                       ([1210], [2])             
Status                                      ([5009], [1])             
Video                                      ([21436], [1])             

       Lifetime Post reach by people who like your Page  \
Type                                                      
Link                                      ([1180], [1])   
Photo                                      ([690], [2])   
Status                                    ([2410], [1])   
Video                                     ([9568], [1])   

       Lifetime People who have liked your Page and engaged with your post  \
Type                                                                         
Link                                          ([19], [1])                    
Photo                                        ([403], [5])                    
Status                                       ([101], [1])                    
Video                                        ([363], [1])                    

            comment           like           share Total Interactions  
Type                                                                   
Link     ([0], [7])  ([12.0], [2])    ([2.0], [3])        ([14], [2])  
Photo   ([0], [94])   ([7.0], [6])  ([14.0], [17])         ([0], [6])  
Status   ([2], [6])  ([86.0], [2])    ([1.0], [3])       ([117], [2])  
Video    ([2], [2])  ([65.0], [1])   ([13.0], [2])        ([81], [1])

In [8]:
#for searching most popular object in dataset I was using following parameters:
#total interaction as main, cause amount of comments + shares is very important, likes - too, but with less weight
#and also reach and impressions for cheching amount of views and demonstrations of posts

df.sort_values(['Total Interactions', 'Lifetime Post Total Reach', 'Lifetime Post Total Impressions'], ascending = [False, False, False]).head(1)

Page total likes   Type  Category  Post Month  Post Weekday  Post Hour  \
244            130791  Photo         2           7             3          5   

     Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
244   1.0                     180480                           319133   

     Lifetime Engaged Users  Lifetime Post Consumers  \
244                    8072                     4010   

     Lifetime Post Consumptions  \
244                        6242   

     Lifetime Post Impressions by people who have liked your Page  \
244                                             108752              

     Lifetime Post reach by people who like your Page  \
244                                             51456   

     Lifetime People who have liked your Page and engaged with your post  \
244                                               3316                     

     comment    like  share  Total Interactions  
244      372  5172.0  790.0                6334

#  

### As u see, winner is the photo #244 in the main dataset
# Thank you for attention